File: 016-trackbar/python/02-trackbar_threshold.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

import cv2

def on_change(x):
    print(x)

origin = cv2.imread('./resources/panad.jpg')
cv2.imshow('origin', origin)

print(origin.shape)

# 添加新窗口
cv2.namedWindow('win_threshold')
# 创建滑块,注册回调函数 lambda x: None没有滑动时
threshold_value = 127
cv2.createTrackbar('color', 'win_threshold', threshold_value, 255, on_change)

while (True):
    threshold_value = cv2.getTrackbarPos('color', 'win_threshold')
    ret, threshold_img = cv2.threshold(origin, threshold_value, 255, cv2.THRESH_BINARY)
    cv2.imshow('win_threshold', threshold_img)

    k = cv2.waitKey(1) & 0xFF
    if (k == 27) or (k == ord('q') ) :
        break

print(threshold_img.shape)

cv2.destroyAllWindows()

File: 016-trackbar/cpp/02_trackbar_threshold.cpp
#include<opencv2/opencv.hpp>
#include<iostream>

using namespace cv;
void on_change(int x, void* ) {
    std::cout<<x<<std::endl;
}

int main(int argc, char *argv[]) {
    Mat mat = imread(argc>1?argv[1]:"./resources/panad.jpg", IMREAD_GRAYSCALE);
    if(mat.empty() ) {
        return -1;
    }
  
    std::cout<<"width/cols="<<mat.cols
        <<" height/rows="<<mat.rows
		<<" channels="<<mat.channels()
		<<std::endl;
    
    std::string winname_color = "origin_color";
    imshow(winname_color, mat);

    Mat dst;
    mat.copyTo(dst);
    std::string threshold_winname= "threshold_color";
    imshow(threshold_winname, dst);
    int color = 100;
    std::string trackbar_name = "color";
    createTrackbar(trackbar_name,  threshold_winname, &color, 255, on_change);


    while(true) {
        int color= getTrackbarPos(trackbar_name, threshold_winname);
        
        threshold(mat, dst, color, 255, THRESH_BINARY);
        imshow(threshold_winname, dst);

        if (waitKey(1) == 'q') 
            break;
    }

    waitKey(0);
    destroyAllWindows();

    return 0;
}
